### PR TITLE
Feature/more string methods

### PR DESCRIPTION
**Pull request by briandowns**
### What's Changed:

Adds collapseSpaces and wrap methods to string type.

### Type of Change:

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#

### Housekeeping:

- [x] Tests have been updated to reflect the changes done within this PR (if applicable).
- [x] Documentation has been updated to reflect the changes done within this PR (if applicable).

#

### CODE REVIEW

**Comment by Jason2605** [comment]
Apologies I will get around to this! Will try my best to have a look tomorrow for you mate

**Comment by briandowns** [comment]
> Apologies I will get around to this! Will try my best to have a look tomorrow for you mate

At your leisure. No rush. :D

File: src/vm/datatypes/strings.c [hunk 1]
```diff
@@ -668,6 +668,51 @@ static Value isLowerString(DictuVM *vm, int argCount, Value *args) {
...
+        runtimeError(vm, "collapseSpaces() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    char *string = AS_CSTRING(args[0]);
+
+    int i, j;
+    for (i = j = 0; string[i]; ++i) {
+        if(!isspace(string[i]) || (i > 0 && !isspace(string[i-1]))) {
+            string[j++] = string[i];
```
Comment (Jason2605) [inline src/vm/datatypes/strings.c:682]: This is unfortunately going to mutate the original string passed in - we'd need to make a copy first.

As a test, run the following code:

```
const x = "hello      test";
x.collapseSpaces();
print(x);
```

File: docs/docs/strings.md [hunk 2]
```diff
@@ -310,3 +310,19 @@ Returns the number of words in the given string. Empty strings are considered fa
...
+
+### string.collapseSpaces() -> String
+
+Returns a string with extraneous whitespace removed.
+
+```cs
+"This  is      a huge string of a       lot of spaces.".collapseSpaces(); // "This is a huge string of a lot of spaces."
+```
+
+### string.wrap() -> String
```
Comment (Jason2605) [inline docs/docs/strings.md:322]:
```suggestion
### string.wrap(Number) -> String
```

File: src/vm/datatypes/strings.c [hunk 3]
```diff
@@ -668,6 +668,51 @@ static Value isLowerString(DictuVM *vm, int argCount, Value *args) {
...
+    int last = 0;
+    int count = 0;
+
+    for (int cur = 0; string[cur] != '\0'; cur++, count++) {
+        if (isspace(string[cur])) {
+            last = cur;
+        }
+
+        if (count >= len) {
+            string[last] = '\n';
```
Comment (Jason2605) [inline src/vm/datatypes/strings.c:708]: Same mutation issue here

File: tests/strings/collapseSpaces.du [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * .collapseSpaces() returns a string with extraneous spaces removed.
+ */
+from UnitTest import UnitTest;
+
+class TestStringCollapseSpaces < UnitTest {
+    testStringCollapseSpaces() {
+        const testString = "This  is      a huge string of a       lot of spaces.";
+        const expected  = "This is a huge string of a lot of spaces.";
+        const res = testString.collapseSpaces();
```
Comment (Jason2605) [inline tests/strings/collapseSpaces.du:14]: Probably worth adding some tests to ensure the original string is also untouched

**Comment by Jason2605** [comment]
Thanks for this!

**Pull request merged**